[pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47818 [pytorch][codegen] migrate gen_variable_factories.py to the new data model
* #47747 [pytorch][codegen] add fully migrated scripts to mypy strict config
* #47746 [pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly
* **#47745 [pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model**

This is a relatively small codegen. Reintroduced 'simple_type' to preserve
old codegen output.

It depends on some methods defined in gen_python_functions.py - next PR will
clean up the remaining Declarations.yaml methods in gen_python_functions.py.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885068](https://our.internmc.facebook.com/intern/diff/D24885068)

ljk53 committed:
[pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model"


This is a relatively small codegen. Reintroduced 'simple_type' to preserve
old codegen output.

It depends on some methods defined in gen_python_functions.py - next PR will
clean up the remaining Declarations.yaml methods in gen_python_functions.py.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885068](https://our.internmc.facebook.com/intern/diff/D24885068)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model"


This is a relatively small codegen. Reintroduced 'simple_type' to preserve
old codegen output.

It depends on some methods defined in gen_python_functions.py - next PR will
clean up the remaining Declarations.yaml methods in gen_python_functions.py.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885068](https://our.internmc.facebook.com/intern/diff/D24885068)

[ghstack-poisoned]

ezyang commented:
So, is the intention to eventually delete `simple_type` again at the end, after more is migrated?

ezyang commented on tools/codegen/api/python.py:
```diff
@@ -447,7 +447,10 @@ def has_tensor_options(f: NativeFunction) -> bool:
-def argument_type_str(t: Type) -> str:
+# 'simple_type' was introduced by the old codegen, which is slightly
+# different from the python schema type, e.g.: doesn't have '?' suffix
+# for optional Tensor/TensorList; doesn't have '[size]' suffix for list type.
```
If `simple_type` is gonna be temporary would be nice to have some explanation of how we can kill it later

ljk53 replied:
@ezyang simple_type is in the output file, e.g.:
```
annotated_args = {
    torch._C._VariableFunctions._cast_Byte: [{'name': 'self', 'simple_type': 'Tensor'}],
```

It's referenced in test/test_overrides.py - if it's the only use case then we can fix later. Not sure if there are other uses.

ezyang replied:
The test/test_overrides.py use is definitely incidental. What they needed there was an easy way to tell if an argument would accept overrides or not, and I think they just implemented it via `simple_type` because it was easier that way. But no intrinsic relationship there.

ezyang commented on tools/codegen/api/python.py:
```diff
@@ -465,31 +468,48 @@ def argument_type_str(t: Type) -> str:
-        elem = argument_type_str(t.elem)
+        if str(t.elem) == 'Tensor':
+            if not simple_type or local.use_c10_dispatcher().dispatcher_uses_new_style():
+                # Is it desired to keep '?' for simple_type with new style dispatcher?
```
Yeah, this condition looks pretty questionable. cc @smessmer

ezyang commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -13,65 +13,69 @@
+    )
+    annotated_args: List[str] = []
+    for pred, namespace in mappings:
+        groups = defaultdict(list)
```
in strict mypy you gonna need a type annotation here

ljk53 replied:
@ezyang hmm, I already added this file to mypy-strict.ini in the sequent PR on this stack and it gave clean result:
```
mypy --config mypy-strict.ini 
Success: no issues found in 11 source files
```
Is there any problem with the config we should fix?

ezyang replied:
Hmm, this will be good to figure out.  When I apply this diff:

```
diff --git a/tools/codegen/gen.py b/tools/codegen/gen.py
index 5324246d6c..5cd4a6ffa8 100644
--- a/tools/codegen/gen.py
+++ b/tools/codegen/gen.py
@@ -1191,5 +1191,7 @@ def main() -> None:
         core_fm.write_outputs(f"{options.output_dependencies}-core")
         cuda_fm.write_outputs(f"{options.output_dependencies}-cuda")
 
+x = defaultdict(list)
+
 if __name__ == '__main__':
     main()
```

and run type checking I get

```
(/home/ezyang/local/pytorch-tmp-env) [ezyang@devvm066.ash0 ~/local/pytorch-tmp] mypy --config mypy-strict.ini 
tools/codegen/gen.py:1194:1: error: Need type annotation for 'x'
Found 1 error in 1 file (checked 7 source files)
```

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model"


This is a relatively small codegen. Reintroduced 'simple_type' to preserve
old codegen output.

It depends on some methods defined in gen_python_functions.py - next PR will
clean up the remaining Declarations.yaml methods in gen_python_functions.py.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885068](https://our.internmc.facebook.com/intern/diff/D24885068)

[ghstack-poisoned]

bhosmer commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -13,65 +13,69 @@
+
+    template_path = os.path.join(autograd_dir, 'templates')
+    fm = FileManager(install_dir=out, template_dir=template_path, dry_run=False)
+    fm.write_with_template('annotated_fn_args.py', 'annotated_fn_args.py', lambda: {
```
...so per our thread in #47818, is this gonna say "generated by aten/src/ATen/gen.py"? 😁  

oh also, curious how changes in these comment lines play in the byte-for-byte comparisons

bhosmer reviewed (approved):
LGTM, mod @ezyang's cc of @smessmer on the `simple_type` new dispatcher thing if it's easily resolved, otherwise I'd leave it for later.

ezyang commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -13,65 +13,69 @@
+                annotated_args.append(f'{namespace}.{process_func(f)}')
+
+    template_path = os.path.join(autograd_dir, 'templates')
+    fm = FileManager(install_dir=out, template_dir=template_path, dry_run=False)
```
A small caution about using FileManager here: there really is only one reason FileManager exists and that's to let us write out the set of files that are going to be generated. So without that connective tissue here there isn't that much point in using FileManager XD

ljk53 replied:
> A small caution about using FileManager here: there really is only one reason FileManager exists and that's to let us write out the set of files that are going to be generated. So without that connective tissue here there isn't that much point in using FileManager XD

Another useful function provided by FileManager is `_write_if_changed()` - we don't want to update the timestamp of generated file if the content doesn't change.
The alternative is to use the write() method defined in utils.py - but I'll have to migrate that file to have clean mypy output.
I feel consolidating all write function to FileManager isn't necessary bad? :)

ezyang commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -13,65 +13,69 @@
+        for f in native_functions:
+            if not should_generate_py_binding(f) or not pred(f):
+                continue
+            groups[str(f.func.name.name)].append(f)
```
nit: technically you don't need the `str` here, because BaseOperatorName has equality and hashing (being frozen) so you could key on it directly

ljk53 replied:
ok, also changed type annotation to 'Dict[BaseOperatorName, ...]'

ezyang commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -13,65 +13,69 @@
+    })
+
+@with_native_function
+def process_func(f: NativeFunction) -> str:
```
I cannot easily tell from the output type (str) or the function name (func) what this function does.

ljk53 replied:
renamed to gen_annotated_args

ezyang commented on tools/autograd/gen_annotated_fn_args.py:
```diff
@@ -13,65 +13,69 @@
-    for e in d.values():
-        for i in e:
-            yield i
+    return f'{f.func.name.name}: {repr(out_args)},'
```
Oh wait, the func.names here have namespaces attached to them?

ljk53 replied:
> Oh wait, the func.names here have namespaces attached to them?

hmm, I'm not sure I understand? you mean 'aten::' namespace or 'torch._C._VariableFunctions' namespace?
We don't have the former in the output file. The latter is added explicitly above.

ezyang reviewed (approved):
okey dokey

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model"


This is a relatively small codegen. Reintroduced 'simple_type' to preserve
old codegen output.

It depends on some methods defined in gen_python_functions.py - next PR will
clean up the remaining Declarations.yaml methods in gen_python_functions.py.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885068](https://our.internmc.facebook.com/intern/diff/D24885068)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model"


This is a relatively small codegen. Reintroduced 'simple_type' to preserve
old codegen output.

It depends on some methods defined in gen_python_functions.py - next PR will
clean up the remaining Declarations.yaml methods in gen_python_functions.py.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885068](https://our.internmc.facebook.com/intern/diff/D24885068)

[ghstack-poisoned]
